lib/progressTracker.js

```diff
@@ -24,26 +24,45 @@ function track (instance, opts) {
   opts = xtend(defaults, opts)
 
   const chalk = new Chalk.constructor({ enabled: testColorSupport({ stream: opts.outputStream }) })
-
   // this default needs to be set after chalk is setup, because chalk is now local to this func
   opts.progressBarString = opts.progressBarString || `${chalk.green('running')} [:bar] :percent`
 
   const iOpts = instance.opts
-  if (opts.renderProgressBar) {
-    let msg = `${iOpts.connections} connections`
+  let durationProgressBar
+  let amountProgressBar
 
-    if (iOpts.pipelining > 1) {
-      msg += ` with ${iOpts.pipelining} pipelining factor`
-    }
+  instance.on('start', () => {
+    if (opts.renderProgressBar) {
+      let msg = `${iOpts.connections} connections`
+
+      if (iOpts.pipelining > 1) {
+        msg += ` with ${iOpts.pipelining} pipelining factor`
+      }
 
-    if (!iOpts.amount) {
-      logToStream(`Running ${iOpts.duration}s test @ ${iOpts.url}\n${msg}\n`)
+      if (!iOpts.amount) {
+        logToStream(`Running ${iOpts.duration}s test @ ${iOpts.url}\n${msg}\n`)
 
-      trackDuration(instance, opts, iOpts)
-    } else {
-      logToStream(`Running ${iOpts.amount} requests test @ ${iOpts.url}\n${msg}\n`)
+        durationProgressBar = trackDuration(instance, opts, iOpts)
+      } else {
+        logToStream(`Running ${iOpts.amount} requests test @ ${iOpts.url}\n${msg}\n`)
+
+        amountProgressBar = trackAmount(instance, opts, iOpts)
+      }
+    }
+  })
 
-      trackAmount(instance, opts, iOpts)
+  // add listeners for progress bar to instance here so they aren't
+  // added on restarting, causing listener leaks
+  if (opts.renderProgressBar && opts.outputStream.isTTY) {
+    if (!iOpts.amount) { // duration progress bar
+      instance.on('tick', () => { durationProgressBar.tick() })
+      instance.on('done', () => { durationProgressBar.tick(iOpts.duration - 1) })
+      process.once('SIGINT', () => { durationProgressBar.tick(iOpts.duration - 1) })
+    } else { // amount progress bar
+      instance.on('response', () => { amountProgressBar.tick() })
+      instance.on('reqError', () => { amountProgressBar.tick() })
+      instance.on('done', () => { amountProgressBar.tick(iOpts.amount - 1) })
+      process.once('SIGINT', () => { amountProgressBar.tick(iOpts.amount - 1) })
```

Comment:
I mean, the code above for both `amountProgressBar` and `durationProgressBar` looks _very_ similar. we should probably factor it out in its own function, ideally calling the same function with different parameters, or something similar.

(this is a nit of style, nothing more)